python/gstaichi/lang/util.py

```diff
@@ -239,17 +239,32 @@ def to_gstaichi_type(dt):
     raise AssertionError(f"Unknown type {dt}")
 
 
+class DataTypeCxxWrapper(_ti_core.DataTypeCxx):
+    __slots__ = ('_hash',)
+
+    def __init__(self, dtype: _ti_core.Type):
+        super().__init__(dtype)
+        self._hash = super().__hash__()
+
+    def __hash__(self):
+        return self._hash
```

Comment:
This wrapper is the only part that is "dirty". It is used to speed up hashing (for now), and could be extended to shape and element_type.